### ISSUE TITLE
The specific limitations of JSON Schema?

### ISSUE THREAD

**Issue by ckhfor** [mlc-ai/xgrammar#104]
Good Job! Can you provide the specific limitations of JSON Schema? Like, additionalProperties,defs,min/max,patternProperties. Thanks.

**Comment by Ubospica** [comment]
Yes. We will update our doc to cover the supported and unsupported keywords soon.

**Comment by gryffindor-rr** [comment]
> Yes. We will update our doc to cover the supported and unsupported keywords soon.

is the doc updated? i'd like to evaluate if xgrammar can meet my required json schema. Thanks!

**Comment by CDboyOne** [comment]
Hello, any update on JSON Schema document?

**Comment by percymcn** [comment]
Measured against **xgrammar 0.2.5** while auditing schema→grammar conversion, in case it helps shape the doc: there is a class of silent failure here that **a keyword-by-keyword table cannot express**, because the fate of a keyword depends on the *node it sits on* rather than on the keyword.

A node that declares a container type but gives no **shape** for it — an object with no `properties`/`additionalProperties`, an array with no `items`/`prefixItems` — compiles to *the empty container only* as soon as almost any further keyword is present. No error, no warning.

```python
import json, xgrammar as xgr
from xgrammar.testing import _is_grammar_accept_string as acc
g = lambda s: xgr.Grammar.from_json_schema(json.dumps(s))

acc(g({"type":"object"}),                      '{"a":1}')  # True   correct
acc(g({"type":"object","required":["a"]}),     '{"a":1}')  # False  <-- 
acc(g({"type":"object","required":["a"]}),     '{}')       # True   <--
```

`root ::= (("{" "}") | ("{" [ \n\t]* "}"))` — the empty object is the only legal document.

`required` is the sharpest case because the resulting accept set is **disjoint** from the schema's: it permits exactly the one document the schema forbids and rejects every document it requires. The array analogue is `contains`, for the same reason.

### Why a keyword list can't capture it

Two controls decide the shape of any doc entry:

| node | `{}` | `{"a":1}` | |
|---|---|---|---|
| `{"type":"object"}` | ✓ | ✓ | correct |
| `{"type":"object","required":["a"]}` | ✓ | ✗ | destroyed |
| `{"type":"object","frobnicate":1}` | ✓ | ✗ | **a keyword you don't implement does it too** |
| `{"type":"object","title":"t"}` | ✓ | ✓ | annotations stay inert |
| `{"type":"object","properties":{"a":{}},"uniqueItems":true}` | ✓ | ✓ | **same keyword, shaped node → harmless** |

The `frobnicate` row is the important one: the trigger is the *absence of a shape*, not membership in any keyword set, so "these keywords are unsupported" would be the wrong shape of statement. The last row is why: the identical keyword on a node that has a shape is byte-identically ignored, which is the documented and expected behaviour.

Arrays behave the same way, and split two ways — worth separating in a doc, since one is loud and one is not:

```python
acc(g({"type":"array","uniqueItems":True}), '["x"]')   # False  (silent)
acc(g({"type":"array","contains":{"type":"string"}}), '["x"]')  # False  (silent)
g({"type":"array","minItems":1})                       # RuntimeError  (loud)
```

### Three details that would otherwise be easy to miss

**1. `type` has two spellings and both collapse.** JSON Schema defines `type` as *"a string or an array of strings"*, and the list form behaves identically:

```python
acc(g({"type":["object","null"],"required":["a"]}), '{"a":1}')  # False
```

This matters more than it looks: `type: ["object","null"]` is the canonical 2020-12 / OpenAPI 3.1 way to say *nullable*, so it is what a lot of real generator output looks like. A fix or a test written against `"type":"object"` alone would miss it. A union naming **both** containers (`["object","array"]`) collapses both halves independently.

**2. `unevaluatedProperties` counts as a shape by its VALUE, not by its presence:**

```python
acc(g({"type":"object","unevaluatedProperties":True, "frobnicate":1}), '{"a":1}')  # True
acc(g({"type":"object","unevaluatedProperties":{},   "frobnicate":1}), '{"a":1}')  # True
acc(g({"type":"object","unevaluatedProperties":False,"frobnicate":1}), '{"a":1}')  # False
```

**3. Test-design warning, since this is the part that makes it survive.** Any fixture whose object declares `properties` — or whose array declares `items` — **passes either way**, because that node has a shape and the keyword is then correctly ignored. So a regression test built from a realistic schema will be green while the defect ships. The discriminating fixture is a container with *no* shape key plus one further keyword, and the assertion has to be that a **valid** document is *accepted* — compiling successfully is precisely what goes wrong.

Happy to open a separate issue for the bare-container behaviour itself if that's more useful than folding it into the doc — I didn't want to file a second one while [#826](https://github.com/mlc-ai/xgrammar/issues/826) is still open.
